### PR TITLE
fix amd gpu memory negative size

### PR DESCRIPTION
**Pull request by neoedmund**
AMD GPU memory negative size. (workaround)

real memory: 
16GB(RX 6800) + 64GB(shared, system has 128GB)

before:
`GPU memory  : -49938MB (total) / -49938MB (available)`

after:
`GPU memory  : 81111MB (total) / 81111MB (available)`

test on Linux
```
	SDL version : 2.0.22 (linked) / 2.0.22 (compiled)
	GL version  : 4.6 (Compatibility Profile) Mesa 20.3.5
	GL vendor   : AMD
	GL renderer : AMD Radeon RX 6800 (SIENNA_CICHLID, DRM 3.46.0, 5.18.0-rc3-gs, LLVM 11.0.1)
	GLSL version: 4.60
	GLEW version: 2.1.0
	GPU memory  : 81031MB (total) / 81031MB (available)
	SDL swap-int: -1
	SDL driver  : x11
	
	Initialized OpenGL Context: 4.6 (Compat)
```

test on Windows (not effected)
```
	SDL version : 2.0.18 (linked) / 2.0.18 (compiled)
	GL version  : 3.0.14800 Compatibility Profile/Debug Context 22.3.2 30.0.15019.1005
	GL vendor   : ATI Technologies Inc.
	GL renderer : AMD Radeon RX 6800
	GLSL version: 4.60
	GLEW version: 2.1.0
	GPU memory  : unknown
	SDL swap-int: 0
	SDL driver  : windows
	
	Initialized OpenGL Context: 3.0 (Compat)
```

ce784e7dd2

### CODE REVIEW

File: rts/Rendering/GL/myGL.cpp [hunk 1]
```diff
@@ -135,7 +135,7 @@ static bool GetVideoMemInfoATI(GLint* memInfo)
 	// these are not disjoint, don't sum
 	for (uint32_t param: {/*GL_VBO_FREE_MEMORY_ATI,*/ GL_TEXTURE_FREE_MEMORY_ATI/*, GL_RENDERBUFFER_FREE_MEMORY_ATI*/}) {
 		glGetIntegerv(param, &memInfo[0]);
-
+		if (memInfo[2] < 0) memInfo[2] = ( 1 << 27 ) + memInfo[2]; // workaround
```
Comment (badosu) [inline rts/Rendering/GL/myGL.cpp:138]: Can you add a comment this is a workaround for amd gpus that might return negative sizes?

Comment (neoedmund) [reply rts/Rendering/GL/myGL.cpp]: Sure, is it looks okay now?

Comment (badosu) [reply rts/Rendering/GL/myGL.cpp]: Yes, thanks!

**Comment by lhog** [comment]
Modern AMD cards/drivers have no means to query for available GPU RAM. You may well return some const value instead.

**Comment by badosu** [comment]
It seems to work on Linux?

**Comment by neoedmund** [comment]
> Modern AMD cards/drivers have no means to query for available GPU RAM

Vulkan can

**Comment by neoedmund** [comment]
> It seems to work on Linux?

Yes, `memInfo[0]`returns 16GB, correct for RX6800.

**Comment by lhog** [comment]
Yeah on Linux it might work, on Windows this extension is missing since 2014: https://community.amd.com/t5/archives-discussions/glgetintegerv-gl-texture-free-memory-ati-param-not-working-on/td-p/115541

> Vulkan can

That is true, but pretty useless for spring as we don't initialize Vulkan.

**Comment by lhog** [comment]
Can you explain what `1 << 27` addition is supposed to do?

**Comment by neoedmund** [comment]
> Can you explain what `1 << 27` addition is supposed to do?

sorry, this is a quick fix to be correct on my machine, which could be wrong on other situation. At least I didn't dig into where the wrong value comes from the upstream to make a perfect fix.

**Comment by lhog** [comment]
N2S. TODO: examine if we can borrow something from `glxinfo` https://cgit.freedesktop.org/mesa/demos/tree/src/xdemos/glinfo_common.c#n913 (unlikely so)

**Comment by lhog** [comment]
I checked and found nothing we don't do already in `glxinfo`, therefore it's advisable to look for other testers with different amounts of VRAM and RAM such that we could verify if this change makes sense.

Also have I got you right that `GL_TEXTURE_FREE_MEMORY_ATI` returns to you 
`16G` at [0] - total memory free in the pool
and 
`-1` at [2] - total auxiliary memory free ?

**Comment by neoedmund** [comment]
[0] returns 16GB
 [2] returns a negative value but not -1 , something -64GB

**Comment by badosu** [comment]
Just for reference AMD returns an accurate value for me on current master, so it's not all GPUs:

```
[t=00:00:00.138305] 	SDL version : 2.0.20 (linked) / 2.0.8 (compiled)
[t=00:00:00.138311] 	GL version  : 4.6 (Compatibility Profile) Mesa 21.3.8
[t=00:00:00.138317] 	GL vendor   : AMD
[t=00:00:00.138323] 	GL renderer : AMD Radeon RX 580 Series (POLARIS10, DRM 3.44.0, 5.16.18-1-MANJARO, LLVM 13.0.1)
[t=00:00:00.138329] 	GLSL version: 4.60
[t=00:00:00.138335] 	GLEW version: 2.2.0
[t=00:00:00.138340] 	GPU memory  : 15212MB (total) / 15212MB (available)
[t=00:00:00.138391] 	SDL swap-int: -1
[t=00:00:00.138398] 	SDL driver  : x11
```

**Pull request closed without merging**